keps/sig-cluster-lifecycle/kubeadm/20190424-kubeadm-for-windows.md

```diff
@@ -0,0 +1,293 @@
+
+Next, the script will run "kubeadm join ..." to join the node to the cluster. In this step kubeadm will run preflight checks and proceed with the regular join process.
+
+kubeadm will then restart the kubelet service using flags that kubeadm fed to the Windows service. kubeadm will proceed to bootstrap the node. After this process is finished the node should show with the status of NotReady.
+
+The same script will then configure FlannelD. The script will (re-)register FlannelD as a service and start it with the correct configuration, optionally using parameters from kubeadm.
+
+kube-proxy will do the same steps as FlannelD shortly after.
+
+After these steps, the node status should be Ready.
+
+### Implementation Details/Notes/Constraints
+
+#### Kubeadm manages the kubelet start / stop as a service
+
+Kubeadm has a “kubelet” phase that can start/stop the [kubelet service](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubelet/kubelet.go).
+
+Kubeadm already supports running the kubelet as a Windows service. This support was added by these two PRs: [60144](https://github.com/kubernetes/kubernetes/pull/60144), [60144](https://github.com/kubernetes/kubernetes/pull/53553).
+
+This existing support will be leveraged by this proposal.
+
+#### Kubeadm makes assumptions about systemd and Linux
+
+Kubeadm is quite coupled to systemd and Linux at the moment.
+This file shows some of the kubelet-flags that are Linux bound:
+[kubelet flags](https://github.com/kubernetes/kubernetes/blob/b8b689aae0d6879de5192b203590330a11c7b9e3/cmd/kubeadm/app/phases/kubelet/flags.go)
+
+The following changes are required::
+* Implement proper cri-socket detection in kubeadm for Windows.
+* Omit flags such as “cgroup-driver”
+* Specify the correct “pod-infra-container-image”
+* Also omit writing systemd drop-in files on Windows in general.
+* Populate the Windows service dynamically.
+
+To support Windows specific flags for the kubelet, there is a requirement to split kubeadm’s app/phases/kubelet/flags.go files into two:
+* app/phases/kubelet/flags_windows.go
+* app/phases/kubelet/flags_linux.go
+
+In the case of Windows, we will be updating the registration of the kubelet as a Windows service, by changing its flags based on the flags kubeadm wants to use, complying with Windows specific defaults in kubeadm.
+
+The existing process of passing the following files to the kubelet, will be leveraged:
+* /etc/kubernetes/bootstrap-kubelet.conf
+* /etc/kubernetes/kubelet.conf
+* /var/lib/kubelet/config.yaml
+
+Windows related adjustments to default paths might be required.
+
+#### Windows vs Linux host paths
+
+Kubeadm makes a number of non-portable assumptions about paths. E.g. “/etc/kubernetes” is a hardcoded path in kubeadm.
+
+We need to evaluate the kubeadm codebase for such instances of non-portable paths - CRI sockets, Cert paths, etc. Such paths need to be defaulted properly in the kubeadm configuration API.
```

Comment:
IMO this should span several releases and be done judiciously, I want us todo this wisely and not in a rush.